xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-default/src/main/java/org/xwiki/notifications/notifiers/internal/UserEventManager.java

```diff
@@ -302,4 +315,30 @@ private boolean isEventFiltered(List<NotificationFilter> filters, Event event, D
         }
         return false;
     }
+
+    public void cleanUpFilters(DocumentDeletedEvent event, DocumentReference user)
+    {
+        DocumentReference deletedDocumentReference = event.getDocumentReference();
+        String serializedReference = this.entityReferenceSerializer.serialize(deletedDocumentReference);
+        try {
+            Set<NotificationFilterPreference> matchingPreferences =
+                this.notificationFilterPreferenceManager.getFilterPreferences(user)
+                    // We only clean up the filters matching exactly the page reference
+                    // we never clean up filters for space / wiki there
+                    .stream().filter(pref -> StringUtils.equals(pref.getPageOnly(), serializedReference))
+                    .collect(Collectors.toSet());
+            if (!matchingPreferences.isEmpty()) {
+                CleaningFilterEvent cleaningFilterEvent = new CleaningFilterEvent();
+                this.observationManager.notify(cleaningFilterEvent, deletedDocumentReference, matchingPreferences);
```

Comment:
I'm not sure what's best between sending a single event with all the preferences about to be deleted, or multiple events for each one. I feel like it's better like this for scalability... 
Also I guess we need to be careful to document that any listener on `CleanFilterEvent` shouldn't perform any operation leading to a delete to not enter in loops...